backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,163 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+const DefaultViewRefreshIntervalSecond = 5
+
+type refreshFunc func(db *gorm.DB)
+
+type ViewRefreshHandler struct {
+	mu                        sync.Mutex
+	viewsToRefresh            map[string][]string // map of tables that shows which views should be refreshed due to table changes
+	tableChanged              map[string]bool
+	refreshFunc               map[string]refreshFunc // map of refresh functions for specified views
+	viewRefreshIntervalSecond time.Duration
+}
+
+func (vh *ViewRefreshHandler) TableChanged(table string) {
+	if !vh.IsSetViewRefreshHandler() {
+		return
+	}
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.tableChanged[table] = true
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	for viewName := range viewNames {
+		vh.refreshFunc[viewName](db)
+	}
+}
+
+// getViewsToRefresh creates a list of views that should be refreshed due to table changes.
+// The viewsToRefresh is the map of tables that shows which views should be refreshed due to table changes.
+func (vh *ViewRefreshHandler) getViewsToRefresh() map[string]bool {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table := range tables {
+		for _, viewName := range vh.viewsToRefresh[table] {
+			viewToRefresh[viewName] = true
+		}
+	}
+
+	return viewToRefresh
+}
+
+func (vh *ViewRefreshHandler) RegisterViewRefreshHandlers(dbDriver string) {
+	switch dbDriver {
+	case DBDriverTypePostgres:
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(applicationViewName),
+			applicationViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(resourceViewName),
+			resourceViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(packageViewName),
+			packageViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(vulnerabilityViewName),
+			vulnerabilityViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+	}
+}
```

Comment:
This still doesn't put this close to the actual view creation and management, I would rather this was a function called alongside initPostgresMaterializedViews like "registerPostgresViewRefreshFunctions". Its not the responsiblity of the ViewRefreshHandler to know anything about the DB driver.

Also if you use `...string` in the function definition for registerViewRefreshHandler there is no need for the extra `[]string{...}` here.